publisher/lock/s3_lock.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2021 New Relic Corporation. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+package lock
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials/stscreds"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+const (
+	defaultTTL = 1000 * time.Hour // disable to manage leftover lockfiles manually for now
+)
+
+// We should parametrise these:
+var (
+	// resource tags
+	tagOwningTeam = "CAOS"
+	tagProduct    = "integrations"
+	tagProject    = "infrastructure-publish-action"
+	tagEnv        = "us-development"
+)
+
+// S3 based lock.
+type S3 struct {
+	client   *s3.S3
+	owner    string
+	bucket   string
+	tags     string
+	filePath string
+	ttl      time.Duration
+}
+
+// lockData represents contents of the JSON lock-file at S3.
+type lockData struct {
+	Owner     string    `json:"owner"`
+	CreatedAt time.Time `json:"created_at"`
+}
+
+// same owner means lock was already acquired by the client
+func (l *lockData) belongsTo(owner string) bool {
+	return l.Owner == owner
+}
+
+func (l *lockData) isExpired(ttl time.Duration, t time.Time) bool {
+	return l.CreatedAt.Add(ttl).Before(t)
+}
+
+// NewS3 creates a lock instance ready to be used validating required AWS credentials.
+func NewS3(bucket, roleARN, region, filepath, owner string) (*S3, error) {
+	sess, err := session.NewSession()
+	if err != nil {
+		return nil, err
+	}
+
+	creds := stscreds.NewCredentials(sess, roleARN, func(p *stscreds.AssumeRoleProvider) {})
+	conf := aws.Config{
+		Credentials: creds,
+		Region:      aws.String(region),
+	}
+
+	return &S3{
+		client:   s3.New(sess, &conf),
+		owner:    owner,
+		bucket:   bucket,
+		filePath: filepath,
+		tags:     fmt.Sprintf("department=product&product=%s&project=%s&owning_team=%s&environment=%s", tagProduct, tagProject, tagOwningTeam, tagEnv),
+		ttl:      defaultTTL,
+	}, nil
+}
+
+// Lock S3 has no compare-and-swap so this is no bulletproof solution, but should be good enough.
+func (l *S3) Lock() error {
+	if l.isBusyDeletingExpired() {
+		return LockBusyErr
```

Comment:
I believe to remember the naming convention for public error is to use `Err` as prefix, am I right?
I couldn't find the go doc about it thou :(

Reply:
you are right, suffix for types, prefix for vars https://github.com/golang/go/wiki/Errors#naming